operators/src/upgrade/controller/reconciler.rs

```diff
@@ -0,0 +1,388 @@
+}
+
+/// Determine what we want to do when dealing with errors from the
+/// reconciliation loop
+fn error_policy(error: &Error, _ctx: Arc<ControllerContext>) -> Action {
+    let duration = Duration::from_secs(match error {
+        Error::Duplicate { timeout } | Error::SpecError { timeout, .. } => (*timeout).into(),
+
+        Error::ReconcileError { .. } => {
+            return Action::await_change();
+        }
+        _ => 5,
+    });
+
+    let when = Utc::now()
+        .checked_add_signed(chrono::Duration::from_std(duration).unwrap())
+        .unwrap();
+    warn!(
+        "{}, retry scheduled @{} ({} seconds from now)",
+        error,
+        when.to_rfc2822(),
+        duration.as_secs()
+    );
+    Action::requeue(duration)
+}
+
+async fn reconcile(ua: Arc<UpgradeAction>, ctx: Arc<ControllerContext>) -> Result<Action, Error> {
+    ensure_crd(ctx.k8s.clone()).await;
+    let ua = ctx.upsert(ctx.clone(), ua).await;
+
+    let _ = ua.finalizer().await;
+
+    match ua.status {
+        Some(UpgradeActionStatus {
+            state: UpgradeState::NotStarted,
+            ..
+        }) => ua.updating().await,
+
+        Some(UpgradeActionStatus {
+            state: UpgradeState::Updating,
+            ..
+        }) => ua.verifying().await,
+
+        Some(UpgradeActionStatus {
+            state: UpgradeState::VerifyingUpdate,
+            ..
+        }) => ua.successfull_update().await,
+
+        Some(UpgradeActionStatus {
+            state: UpgradeState::SuccessfullUpdate,
+            ..
+        }) => ua.delete().await,
+
+        Some(UpgradeActionStatus {
+            state: UpgradeState::Error,
+            ..
+        }) => {
+            error!(upgrade = ?ua.name_any(), "entered error as final state");
+            Err(Error::ReconcileError {
+                name: ua.name_any(),
+            })
+        }
+
+        // We use this state to indicate its a new CRD however, we could (and
+        // perhaps should) use the finalizer callback.
+        None => ua.start().await,
+    }
+}
+
+/// Upgrade controller that has its own resource and controller context.
+pub async fn upgrade_controller(args: &ArgMatches) -> Result<(), Error> {
+    let ua: Api<UpgradeAction> = Api::namespaced(
+        UpgradeOperatorConfig::get_config().k8s_client().client(),
+        UpgradeOperatorConfig::get_config().namespace(),
+    );
+    let lp = ListParams::default();
+
+    let context = ControllerContext {
+        k8s: UpgradeOperatorConfig::get_config().k8s_client().client(),
+        inventory: tokio::sync::RwLock::new(HashMap::new()),
+        interval: args
+            .value_of("interval")
+            .unwrap()
+            .parse::<humantime::Duration>()
+            .expect("interval value is invalid")
+            .as_secs(),
+        retries: args
+            .value_of("retries")
+            .unwrap()
+            .parse::<u32>()
+            .expect("retries value is invalid"),
+    };
+
+    Controller::new(ua, lp)
```

Comment:
How many concurrent reconcilers will this spawn? Will this spawn as a separate tokio task? Controller reconcilers should not be run synchronously.